jetbrains-core/src/software/aws/toolkits/jetbrains/services/s3/objectActions/DeleteObjectAction.kt

```diff
@@ -0,0 +1,91 @@
+// Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+package software.aws.toolkits.jetbrains.services.s3.objectActions
+
+import com.intellij.icons.AllIcons
+import com.intellij.openapi.actionSystem.AnActionEvent
+import com.intellij.openapi.actionSystem.LangDataKeys
+import com.intellij.openapi.application.ApplicationManager
+import com.intellij.openapi.ui.Messages
+import com.intellij.ui.AnActionButton
+import org.jetbrains.annotations.TestOnly
+import software.amazon.awssdk.services.s3.S3Client
+import software.amazon.awssdk.services.s3.model.Delete
+import software.amazon.awssdk.services.s3.model.DeleteObjectsRequest
+import software.amazon.awssdk.services.s3.model.ObjectIdentifier
+import software.aws.toolkits.core.telemetry.TelemetryNamespace
+import software.aws.toolkits.jetbrains.core.AwsClientManager
+import software.aws.toolkits.jetbrains.services.s3.S3VirtualBucket
+import software.aws.toolkits.jetbrains.services.s3.S3VirtualDirectory
+import software.aws.toolkits.jetbrains.services.s3.bucketEditor.S3KeyNode
+import software.aws.toolkits.jetbrains.services.s3.bucketEditor.S3TreeTable
+import software.aws.toolkits.jetbrains.services.telemetry.TelemetryService
+import software.aws.toolkits.jetbrains.utils.notifyInfo
+import software.aws.toolkits.resources.message
+import javax.swing.tree.DefaultMutableTreeNode
+
+class DeleteObjectAction(
+    private var treeTable: S3TreeTable,
+    val bucket: S3VirtualBucket
+
+) : AnActionButton("Delete Object", null, AllIcons.Actions.Cancel), TelemetryNamespace {
+
+    @Suppress("unused")
+    override fun actionPerformed(e: AnActionEvent) {
+        val project = e.getRequiredData(LangDataKeys.PROJECT)
+        val client: S3Client = AwsClientManager.getInstance(project).getClient()
+        val rows = treeTable.selectedRows
+        val objectsToDelete = mutableListOf<ObjectIdentifier>()
+
+        for (row in rows) {
+            val path = treeTable.tree.getPathForRow(row)
+            val node = (path.lastPathComponent as DefaultMutableTreeNode).userObject as S3KeyNode
+            val file = node.virtualFile
+            val key = when (file.parent is S3VirtualDirectory) {
+                true -> "${file.parent.name}/${file.name}"
+                false -> file.name
+            }
+            objectsToDelete.add(ObjectIdentifier.builder().key(key).build())
+        }
+
+        val response = Messages.showOkCancelDialog(
+            project,
+            "Are you sure you want to delete ${rows.size} objects?",
+            message("delete_resource.title", "Object", ""),
+            "Delete",
+            "Cancel", Messages.getWarningIcon()
+        )
+        if (response == 0) {
+            ApplicationManager.getApplication().executeOnPooledThread {
+                try {
+                    deleteObjectAction(client, objectsToDelete)
+                    TelemetryService.getInstance().record(e.project, "s3") {
```

Comment:
All these telemetry report should go to some action wrapper so we don't have to deal with it all the time.